xapian-letor/ranker/ranker.cc

```diff
@@ -192,35 +192,31 @@ write_to_file(const std::vector<Xapian::FeatureVector> & list_fvecs, const strin
     }
 }
 
+// Query file is in the format: <qid> '<query_string>'
 static std::pair<string, string>
 parse_query_string(const string & query_line, int line_number)
 {
-    vector<string> token;
-    size_t i = 0;
-    size_t j = query_line.find_first_of(' ', i);
-    token.push_back(query_line.substr(i, j - i));
+    string::size_type i = 0;
+    string::size_type j = query_line.find_first_of(' ', i);
+    string qid = query_line.substr(i, j - i);
+    string querystr;
     if (j != string::npos) {
-	i = query_line.find_first_of("'", j);
-	j = string::npos;
-	if (j > i) {
-	    token.push_back(query_line.substr(i, j - i));
+	i = query_line.find_first_not_of("' ", j);
```

Comment:
This code is a bit odd - we don't actually check for a `'` at the start of the query string, but instead skip arbitrary numbers of spaces and `'`.

So this code won't complain about cases like:

```
1234 no open quote'
```

And cases like this will be mishandled:

```
1234 ''tis a shame'
```

Either `'` in query string is allowed (in which case this goes wrong because it strips any leading `'` in the query string, or it isn't (in which case this should give an error but doesn't), or `'` needs to be escaped in the query string in some way (in which case we need to handle undoing such escaping, and this case also ought to be an error because the `'` isn't escaped).

The format for this file ideally ought to specify if `'` in the query string is allowed and if it needs escaping and how.  If not, we'll need to decide how to resolve that underspecification.

This code also mishandles cases where the query string is empty or consists only of spaces, such as:

```
1234 ''
1234 ' '
```

For these, `i` is `string::npos` and it looks to me like that will cause `query_line.substr(i, j - i)` to throw `std::out_of_range` (see https://en.cppreference.com/w/cpp/string/basic_string/substr)

Reply:
> Either `'` in query string is allowed (in which case this goes wrong because it strips any leading `'` in the query string, or it isn't (in which case this should give an error but doesn't), or `'` needs to be escaped in the query string in some way (in which case we need to handle undoing such escaping, and this case also ought to be an error because the `'` isn't escaped).
> 
I think it should be allowed. The fixup commit handles all the cases you mentioned and also we won't need to undo the escaping.

Reply:
OK.

It would be good to add the cases which were wrongly handled to serve as regression test cases.

For the cases which were wrongly failing, we can just add them to a query file and check it parses without an exception.

For the cases which were wrongly being accepted we can just test that `LetorParseError` is thrown.

The `1234 ''tis a shame'` case seems harder as QueryParser parses that the same without the leading `'`.

Reply:
sure will add that in a fixup commit.